File: docs/test-cases/path-traversal.go
// Path traversal — intentionally vulnerable. DO NOT deploy.
package main

import (
	"net/http"
	"os"
	"path/filepath"
)

const uploadDir = "/app/uploads"

// BUG: filepath.Join does NOT prevent traversal — ../../etc/passwd works
func downloadHandler(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")
	path := filepath.Join(uploadDir, filename)
	data, _ := os.ReadFile(path)
	w.Write(data)
}

// BUG: no containment check after join
func readHandler(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	path := filepath.Join(uploadDir, name)
	// filepath.Join cleans ".." but doesn't verify result is under uploadDir
	http.ServeFile(w, r, path)
}

func main() {
	http.HandleFunc("/download", downloadHandler)
	http.HandleFunc("/read", readHandler)
	http.ListenAndServe(":8080", nil)
}

File: docs/test-cases/path-traversal.rs
// Path traversal — intentionally vulnerable. DO NOT deploy.
use actix_web::{get, web, App, HttpResponse, HttpServer};
use std::fs;

const UPLOAD_DIR: &str = "/app/uploads";

// BUG: format! with user input allows traversal
#[get("/download")]
async fn download(query: web::Query<FileQuery>) -> HttpResponse {
    let path = format!("{}/{}", UPLOAD_DIR, query.file);
    // No canonicalization — ../../etc/passwd works
    match fs::read(&path) {
        Ok(data) => HttpResponse::Ok().body(data),
        Err(e) => HttpResponse::NotFound().body(e.to_string()),
    }
}

// BUG: Path::join doesn't prevent absolute paths on all platforms
#[get("/read")]
async fn read_file(query: web::Query<FileQuery>) -> HttpResponse {
    let path = std::path::Path::new(UPLOAD_DIR).join(&query.file);
    // No containment check after join
    match fs::read_to_string(&path) {
        Ok(content) => HttpResponse::Ok().body(content),
        Err(e) => HttpResponse::InternalServerError().body(e.to_string()),
    }
}

#[derive(serde::Deserialize)]
struct FileQuery {
    file: String,
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| App::new().service(download).service(read_file))
        .bind("0.0.0.0:8080")?
        .run()
        .await
}
